lib/modules/dosomething/dosomething_api/resources/magick_resource.inc

```diff
@@ -0,0 +1,144 @@
+        'file' => [
+          'type' => 'inc',
+          'module' => 'dosomething_api',
+          'name' => 'resources/magick_resource',
+        ],
+        'callback' => '_magick_resource_index',
+        'args' => [
+          [
+            'name' => 'action',
+            'description' => 'The action method to access.',
+            'type' => 'string',
+            'optional' => FALSE,
+            'source' => [
+              'param' => 'action',
+            ],
+          ],
+          [
+            'name' => 'campaign',
+            'description' => 'The id of the specified campaign.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'campaign',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'campaign_run',
+            'description' => 'The id of the specified campaign run.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'campaign_run',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'user',
+            'description' => 'The id of the specified user.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'user',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'key',
+            'description' => 'Custom key to allow automatic campaign signup.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'key',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'source',
+            'description' => 'Source for the request.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'source',
+            ],
+            'default value' => NULL,
+          ],
+        ],
+        'access callback' => '_magick_resource_access',
+        'access arguments' => ['index'],
+      ],
+
+    ],
+
+  ];
+
+  return $magick_resource;
+}
+
+/**
+ * Determine whether the method is accessable.
+ *
+ * @return bool
+ */
+function _magick_resource_access($method) {
+  // Temp universal access for now.
+  return TRUE;
+}
+
+/**
+ * Mediate the request and pass along to specfied action if
+ * available.
+ *
+ * @param  string $action
+ * @param  string $campaign
+ * @param  string $user
+ * @return array|ServicesException
+ */
+function _magick_resource_index($action, $campaign, $campaign_run, $user, $key, $source) {
+  if ($action === 'signup') {
+    return magick_signup(compact('campaign', 'campaign_run', 'user', 'key', 'source'));
+  }
+
+  return services_error('This endpoint is specifically for calling custom actions (spells). Please refer to the documentation.', 401);
```

Comment:
Should this be a `400 Bad Request`, since it's a problem with the request (rather than an authentication thing)?